Sources/Rendering/WebGPU/UniformBuffer/index.js

```diff
@@ -0,0 +1,338 @@
+    if (idx === undefined) {
+      vtkErrorMacro(`entry named ${name} not found in UBO`);
+      return;
+    }
+    const entry = model.bufferEntries[idx];
+    publicAPI.createView(entry.nativeType);
+    const view = model[entry.nativeType];
+    if (view[entry.offset / view.BYTES_PER_ELEMENT] !== val) {
+      view[entry.offset / view.BYTES_PER_ELEMENT] = val;
+      model.sendDirty = true;
+    }
+  };
+
+  publicAPI.setArray = (name, arr) => {
+    publicAPI.sortBufferEntries();
+    const idx = model.bufferEntryNames.get(name);
+    if (idx === undefined) {
+      vtkErrorMacro(`entry named ${name} not found in UBO`);
+      return;
+    }
+    const entry = model.bufferEntries[idx];
+    publicAPI.createView(entry.nativeType);
+    const view = model[entry.nativeType];
+    for (let i = 0; i < arr.length; i++) {
+      if (view[entry.offset / view.BYTES_PER_ELEMENT + i] !== arr[i]) {
+        view[entry.offset / view.BYTES_PER_ELEMENT + i] = arr[i];
+        model.sendDirty = true;
+      }
+    }
+  };
+
+  publicAPI.getNativeView = (name) => {
+    publicAPI.sortBufferEntries();
+    const idx = model.bufferEntryNames.get(name);
+    if (idx === undefined) {
+      vtkErrorMacro(`entry named ${name} not found in UBO`);
+      return undefined;
+    }
+    const entry = model.bufferEntries[idx];
+    publicAPI.createView(entry.nativeType);
+    const bpe = model[entry.nativeType].BYTES_PER_ELEMENT;
+    model.sendDirty = true;
+    return new window[entry.nativeType](
+      model.arrayBuffer,
+      entry.offset,
+      entry.sizeInBytes / bpe
+    );
+  };
+
+  publicAPI.getSendTime = () => model.sendTime.getMTime();
+  publicAPI.getShaderCode = () => {
+    // sort the entries
+    publicAPI.sortBufferEntries();
+
+    let result = `[[block]] struct ${model.name}Struct\n{\n`;
+    for (let i = 0; i < model.bufferEntries.length; i++) {
+      const entry = model.bufferEntries[i];
+      result += `[[offset(${entry.offset})]] ${entry.name}: ${entry.type};\n`;
+    }
+    result += `};\n[[binding(${model.binding}), group(${model.group})]] var<uniform> ${model.name}: ${model.name}Struct;`;
+    return result;
+  };
+}
+
+// ----------------------------------------------------------------------------
+// Object factory
+// ----------------------------------------------------------------------------
+
+const DEFAULT_VALUES = {
+  bufferEntries: null,
+  bufferEntryNames: null,
+  sizeInBytes: 0,
+  name: null,
+  binding: 0,
+  group: 0,
+};
+
+// ----------------------------------------------------------------------------
+
+export function extend(publicAPI, model, initialValues = {}) {
+  Object.assign(model, DEFAULT_VALUES, initialValues);
+
+  // Build VTK API
+  macro.obj(publicAPI, model);
+
+  // Internal objects
+  model.bufferEntryNames = new Map();
+  model.bufferEntries = [];
+
+  model.sendTime = {};
+  macro.obj(model.sendTime, { mtime: 0 });
+
+  model.sendDirty = true;
+  model.sortDirty = true;
```

Comment:
should they be part of the defaultValues?

Reply:
Nah, they are the protected flavor of public.